### PR TITLE
Re-implement `list_dir`

### LINKED ISSUE #321: `list_dir` relies on `list_prefix` which traverses all chunks
This is a problem because it's much slower than it needs to be, and it's used by zarr to implement things like `tree`. We should make `list_dir` smarter.

### PR DESCRIPTION
**Pull request by dcherian**
Closes #321

### CODE REVIEW

File: icechunk-python/tests/test_zarr/test_store/test_icechunk_store.py [hunk 1]
```diff
@@ -301,6 +303,12 @@ async def test_set_if_not_exists(self, store: IcechunkStore) -> None:
 
     async def test_is_empty(self, store: IcechunkStore) -> None:
         assert await store.is_empty("")
+        await self.set(
```
Comment (dcherian) [inline icechunk-python/tests/test_zarr/test_store/test_icechunk_store.py:306]: There are no implicit groups in Zarr V3, so I think this is fine. Anyone using the group API would have created these keys.

cc @jhamman

File: icechunk/src/store.rs [hunk 2]
```diff
@@ -428,31 +423,90 @@ impl Store {
...
+                                })
+                            } else {
+                                None
+                            };
+                            Ok(res)
+                        })
+                        .try_collect::<HashSet<_>>()
+                        .await?
+                        .into_iter()
+                        .collect::<Vec<_>>()
```
Comment (dcherian) [inline icechunk/src/store.rs:502]: very yuck, I guess I could just create the stream in each arm of the match instead of returning a Vec

**Comment by dcherian** [comment]
```
FAILED tests/test_can_read_old.py::test_icechunk_can_read_old_repo - AssertionError: assert ['c', 'zarr.json'] == ['zarr.json']
  
  At index 0 diff: 'c' != 'zarr.json'
  Left contains one more item: 'zarr.json'
  
  Full diff:
    [
  +     'c',
        'zarr.json',
    ]
```

I don't understand this, there is a manifest but no chunks were ever written?

File: icechunk/src/store.rs [hunk 3]
```diff
@@ -428,31 +423,90 @@ impl Store {
...
-                // If the prefix is "/", get rid of it. This can happen when prefix is missing
-                // the trailing slash (as it does in zarr-python impl)
-                let rem = &s[idx..].trim_start_matches('/');
-                match rem.split_once('/') {
-                    Some((prefix, _)) => ListDirItem::Prefix(prefix.to_string()),
-                    None => ListDirItem::Key(rem.to_string()),
+        let session = Arc::clone(&self.session).read_owned().await;
+        let absolute_prefix =
+            if !prefix.starts_with("/") { &format!("/{}", prefix) } else { prefix };
+        let prefix = prefix.trim_end_matches("/");
```
Comment (paraseba) [inline icechunk/src/store.rs:429]: should you trim absolute_prefix instead?

File: icechunk/src/store.rs [hunk 4]
```diff
@@ -428,31 +423,90 @@ impl Store {
...
+
+        #[allow(clippy::expect_used)]
+        let path =
+            Path::try_from(absolute_prefix).expect("could not create path from prefix");
```
Comment (paraseba) [inline icechunk/src/store.rs:433]: Let's replace by a not panic error

File: icechunk/src/store.rs [hunk 5]
```diff
@@ -428,31 +423,90 @@ impl Store {
...
+        let results = match session.get_node(&path).await {
+            Ok(NodeSnapshot {
+                id: node_id,
+                node_data: NodeData::Array(.., manifests),
+                ..
+            }) => {
+                // if this is an array we know what to yield
+                let mut res = vec![ListDirItem::Key("zarr.json".to_string())];
+                if session.array_has_chunks_set(&node_id) || !manifests.is_empty() {
```
Comment (paraseba) [inline icechunk/src/store.rs:442]: hmm I don't think this condition is precise enough. It could have chunks in the manifests but all deleted by the session. I think we probably want to call some existing function that lists all chunks for the array and just verify if there are any.

File: icechunk/src/session.rs [hunk 6]
```diff
@@ -317,6 +339,19 @@ impl Session {
         }
     }
 
+    pub async fn array_chunk_iterator<'a>(
+        &'a self,
+        path: &Path,
+    ) -> impl Stream<Item = SessionResult<ChunkInfo>> + 'a {
```
[the remark this answers is not in the document]
Comment (dcherian) [reply icechunk/src/session.rs:345]: apparently not

File: icechunk/src/store.rs (hunk 5 above)
Comment (dcherian) [reply icechunk/src/store.rs:442]: yeah I was trying to be too clever about not downloading a manifest. That can be a future optimization.

File: icechunk/src/store.rs (hunk 2 above)
Comment (dcherian) [reply icechunk/src/store.rs:502]: no that doesn't work becausse HashSet.into_iter() and Vec.into_iter() are different types :/

File: icechunk/src/store.rs (hunk 5 above)
Comment (paraseba) [reply icechunk/src/store.rs]: Ahh I forgot about that! I think this is worth exploring because it can be a big optimization. Maybe if there are manifests and no chunk deletes....? We would have to think about edge cases ... but I like this idea.

File: icechunk-python/tests/test_zarr/test_store/test_icechunk_store.py (hunk 1 above)
Comment (dcherian) [reply icechunk-python/tests/test_zarr/test_store/test_icechunk_store.py:306]: `get_node("")` fails because there a root group was not explicitly created, and we go down the "array chunk list" path. I think this is another case where Icechunk is not a key-value store...

File: icechunk/src/store.rs (hunk 5 above)
Comment (dcherian) [reply icechunk/src/store.rs]: OK I tried again.

Comment (dcherian) [reply icechunk/src/store.rs]: Now I've added a couple more deletion rules to the stateful tests in addition to a list_dir rule to gain more confidence.

Comment (dcherian) [reply icechunk/src/store.rs]: OK I think the one edge-case not addressed is when the user calls `delete_dir("path/to/array/c")`. This is not a deleted array, and we can't detect this without checking that every chunk of the array was deleted or never written in the first place :/

File: icechunk/src/session.rs [hunk 7]
```diff
@@ -240,6 +242,26 @@ impl Session {
             .map(|node| self.change_set.update_array(node.id, metadata))
     }
 
+    pub async fn array_has_modified_chunks(
+        &self,
+        path: Path,
+        id: NodeId,
+    ) -> SessionResult<bool> {
+        let array = self.get_array(&path).await?;
```
Comment (paraseba) [inline icechunk/src/session.rs:250]: I think the caller always has executed get_array already? Maybe we should include in the arguments to this function the `NodeData`? I think with that you can also make this function return `bool`

File: icechunk/src/session.rs [hunk 8]
```diff
@@ -240,6 +242,26 @@ impl Session {
             .map(|node| self.change_set.update_array(node.id, metadata))
     }
 
+    pub async fn array_has_modified_chunks(
```
Comment (paraseba) [inline icechunk/src/session.rs:245]: the name of this function confuses me, because it makes me think this returns true if the session (`self`) has modified any chunks in the array, but I  don't think that's what the function does.

File: icechunk/src/store.rs [hunk 9]
```diff
@@ -428,31 +426,100 @@ impl Store {
...
+                path: node_path,
+                node_data: NodeData::Array(..),
+                ..
+            }) => {
+                // if this is an array we know what to yield
+                let mut res = vec![ListDirItem::Key("zarr.json".to_string())];
+                // The `c` prefix will exist when an array was created and a chunk was written
+                // either (1) in the changeset or (2) as part of a previous snapshot.
+                // It will not exist if (3) the array is recorded as deleted in the change_set.
+                if session.array_has_modified_chunks(node_path, node_id).await? {
```
Comment (paraseba) [inline icechunk/src/store.rs:447]: maybe we are working too hard for something that doesn't really matter? What happens if we always include the `/c`? I don't imagine anything will break.

File: icechunk/src/session.rs [hunk 10]
```diff
@@ -240,6 +242,26 @@ impl Session {
...
+        path: Path,
+        id: NodeId,
+    ) -> SessionResult<bool> {
+        let array = self.get_array(&path).await?;
+        if let NodeSnapshot { node_data: NodeData::Array(_, manifests), .. } = array {
+            debug_assert_eq!(array.id, id);
+            let path_and_id = &(path, id);
+            Ok(!self.change_set.array_is_deleted(path_and_id)
+                && (self.change_set.array_has_modified_chunks(path_and_id)
+                    || !manifests.is_empty()))
```
Comment (paraseba) [inline icechunk/src/session.rs:256]: I think this misses the case of a session that deletes all chunks in an existing array. But see my comment below, maybe we are doing too much... maybe we can always return the `/c`

Comment (dcherian) [reply icechunk/src/session.rs:256]: yes... i had the same realization in https://github.com/earth-mover/icechunk/pull/528#discussion_r1900282037.

I've updated to always return `c`. We can get smarter if someone complains and really cares about it.

File: icechunk/src/session.rs (hunk 8 above)
Comment (dcherian) [reply icechunk/src/session.rs:245]: function is now deleted.

File: icechunk-python/tests/test_zarr/test_stateful.py [hunk 11]
```diff
@@ -107,6 +122,73 @@ def clear(self) -> None:
...
+        note(f"delete_dir with {path=!r}")
+        self._sync(self.model.delete_dir(path))
+        self._sync(self.store.delete_dir(path))
+
+        matches = set()
+        for node in self.all_groups | self.all_arrays:
+            if node.startswith(path):
+                matches.add(node)
+        self.all_groups = self.all_groups - matches
+        self.all_arrays = self.all_arrays - matches
```
[the remark this answers is not in the document]
Comment (dcherian) [reply icechunk-python/tests/test_zarr/test_stateful.py:190]: I added an assert to the `list_prefix` invariant below to make sure the tracked state matches the `store` and `model`

File: icechunk/src/change_set.rs [hunk 12]
```diff
@@ -48,6 +48,15 @@ impl ChangeSet {
         self.updated_attributes.keys()
     }
 
+    pub fn array_is_deleted(&self, path_and_id: &(Path, NodeId)) -> bool {
+        self.deleted_arrays.contains(path_and_id)
+    }
+
+    pub fn array_has_modified_chunks(&self, path_and_id: &(Path, NodeId)) -> bool {
```
Comment (paraseba) [inline icechunk/src/change_set.rs:55]: we can delete this function now right? Just because i don't like the name

File: icechunk/src/store.rs [hunk 13]
```diff
@@ -136,8 +138,9 @@ impl Store {
     }
 
     pub async fn is_empty(&self, prefix: &str) -> StoreResult<bool> {
-        let res = self.list_dir(prefix).await?.next().await;
-        Ok(res.is_none())
+        let list = self.list_dir(prefix).await?;
+        pin_mut!(list);
+        Ok(list.next().await.is_none())
```
Comment (paraseba) [inline icechunk/src/store.rs:143]: is pin still needed now that we didn't change the signature?

**Review by paraseba — APPROVED** [review]
